File: Java/Exercices/workspaces/solid-ocp-end/src/main/java/zenika/solid/ocp/Trip.java
package zenika.solid.ocp;

import static java.util.Objects.requireNonNull;

public class Trip {
    
    private final City departure, arrival;

    private Trip(City departure, City arrival) {
        this.departure = departure;
        this.arrival = arrival;
    }

    public static TripBuilder from(City departure) {
        return new TripBuilder(departure);
    }

    public static class TripBuilder {
        private City departure, arrival;
        
        public TripBuilder(City departure) {
            this.departure = requireNonNull(departure);
        }

        public Trip to(City arrival) {
            this.arrival = requireNonNull(arrival);
            return new Trip(this.departure, this.arrival);
        }
    }
}

File: SOLID/OpenClosePrinciple/Trip.cs
namespace SOLID.OpenClosePrinciple
{

    public class Trip
    {

        private readonly City departure;
        private readonly City arrival;

        private Trip(City departure, City arrival)
        {
            this.departure = departure;
            this.arrival = arrival;
        }

        public static TripBuilder from(City departure)
        {
            return new TripBuilder(departure);
        }

        public class TripBuilder
        {
            private City departure;
            private City arrival;

            public TripBuilder(City departure)
            {
                this.departure = departure;
            }

            public Trip to(City arrival)
            {
                this.arrival = arrival;
                return new Trip(this.departure, this.arrival);
            }
        }
    }
}